graphql_compiler/query_formatting/graphql_formatting.py

```diff
@@ -1,63 +1,60 @@
-        r'\], '
-        r'op_name: '
-        r'("[^"]+")'                    # capture group 2 on this line
-        r'\)'
-    )
-
-    replacement_pattern = (
-        r'@filter(op_name: \2, value: [\1])'
-    )
-
-    return re.sub(filter_directive_pattern, replacement_pattern, query)
-
-
-def pretty_print_graphql(query, use_four_spaces=True):
+def pretty_print_graphql(query, spaces=4):
     """Take a GraphQL query, pretty print it, and return it."""
-    # Use the built-in AST printer to get the canonical representation.
-    output = print_ast(parse(query))
-
-    # Use four spaces for indentation, to make it easier up to edit in Python source files.
-    if use_four_spaces:
-        output = fix_indentation_depth(output)
-
-    # @filter directives are much easier to read if the operation comes before the values.
-    # The default AST printer seems to like outputting the values before the operation, so
-    # the filtering operation ends up written in postfix order. Use regex to correct that.
-    output = fix_filter_directive_order(output)
-
-    return output
+    # The graphql-core printer module has an indent function outside of the
+    # PrintingVisitor class which has two spaces hard coded. Ideally, it would
+    # be a member of the class and we could simply override it in our custom
+    # printer. To work around this we can set the function inside the module
+    # forcefully. This is undesireable as it has a global effect and relies on
+    # the implementation details of the print_ast() function, but has the
+    # advantage of only modifying the leave_Directive() method on the class as
+    # opposed to copying the entire class.
+    import graphql
+
+    def indent_override(maybe_str):
+        if maybe_str:
+            return maybe_str.replace('\n', '\n' + ' ' * spaces)
+        return maybe_str
+    graphql.language.printer.indent = indent_override
+
+    return visit(parse(query), CustomPrintingVisitor())
+
+
+class CustomPrintingVisitor(PrintingVisitor):
+    # Directives are easier to read if their arguments appear in the order in
+    # which we defined them in the schema. For example, @filter directives are
+    # much easier to read if the operation comes before the values. The
+    # arguments of the directives specified in the schema are defined as
+    # OrderedDicts which allows us to sort the provided arguments to match.
+    def leave_Directive(self, node, *args):
+        """Call when exiting a directive node in the ast."""
+        args = node.arguments
+
+        def _arg_index(arg, directive):
+            defined_arg_names = directive.args.keys()
+            # Taking [0] is ok here because the graphql parser checks for the
+            # existence of ':' in directive arguments
+            arg_name = arg.split(':', 1)[0]
+            try:
+                return defined_arg_names.index(arg_name)
+            except ValueError:
```

Comment:
I've always felt conflicted about this myself, but when writing python have aimed to follow the 'easier to ask for forgiveness than permission' guideline. I agree that in this case and many others it reduces the clarity of semantics and intent.